### PR TITLE
curb resource-level access request write capabilities

### PR DESCRIPTION
**Pull request by fspmarshall**
The very first iteration of the access request API had a very simplistic access-control model, where admins could approve any request at any time.  It was mostly intended to be used for break-glass scenarios, and for integration with custom automation.  Since then, we've moved away from the model of blanket permissions being granted to a few, in favor of granular and narrowly-scoped review permissions being granted to many.  In this new model, the old blanket resource-level permissions have become a bit of a footgun.  While we likely can't deprecate them entirely, we can soften some of their rough edges.

This PR introduces two minor breaking changes to resource-level access request write permissions which should slightly reduce their power without meaningfully impacting intended usecases (e.g. custom automation):

- Directly setting states for *your own* access requests is no longer supported.
- Directly setting states for access requests is no longer permitted when the calling identity has active requests applied. 

changelog: blanket/resource-level access request write permissions no longer support self-approval or recursion.

### CODE REVIEW

**Comment by Tener** [comment]
I can see valid use cases for both scenarios. For example this:
```
- Directly setting states for your own access requests is no longer supported.
```
is somewhat equivalent to having low-privilege account for day-to-day work with an occasional `sudo -s` when needed.

This can be very disruptive if you depend on that kind of workflow, with S1 potential. It also isn't clear how widespread these are.

My suggestion:
1. Explicitly document the intended alternatives for both workflows.
2. Make sure to wait for this breaking change at least until the next major release.
3. Consider issuing deprecation warnings for some time before making the actual change.
4. Gather some telemetry to gauge how common are these workflows.

**Comment by zmb3** [comment]
Changes sound reasonable but can we lock this behavior in with a few tests?

**Comment by fspmarshall** [comment]
@Tener you're right that it is possible that some folks use the resource verbs to self-approve.  I'm leaning toward leaving this as-is though.  The main reason is that we've observed that its not uncommon for users to misunderstand the access request resource verbs and think that they need to be granted to users in order for them to be able to use access requests _at all_, rather than being the access request equivalent of unfettered/admin capabilities.  IMO avoiding disruption of unintended and undocumented workflow patterns that don't meaningfully change security is of a lower priority than mitigating the fallout of accidental misuse and misunderstanding of the _intended_ usecase.

Furthermore, I think that while the potential disruption is certainly nonzero, the remedy is also fairly trivial as breaking changes go.  The returned access denied messages are unambiguous, and any roles currently being self-requested just need to move to being statically assigned to return to full function.  Additionally, per-session MFA can play a similar role to the sudo-like workflow you describe, with the bonus of actually being meaningful security-wise.

**Review by rosstimothy — APPROVED** [review]

File: lib/auth/auth_with_roles.go [hunk 1]
```diff
@@ -2503,6 +2503,18 @@ func (a *ServerWithRoles) CreateAccessRequestV2(ctx context.Context, req types.A
...
+	if !req.GetState().IsPending() {
+		if authz.IsCurrentUser(a.context, req.GetUser()) {
+			return nil, trace.AccessDenied("cannot create access request for self in non-pending state")
+		}
+
+		if len(a.context.Identity.GetIdentity().ActiveRequests) != 0 {
+			return nil, trace.AccessDenied("cannot create access requests in non-pending state while using an access request")
+		}
+
+		log.Warnf("Use of resource-level access request 'create' permission by user %q to create non-pending access request for user %q. Creation of non-pending requests will be deprecated in future version of teleport. Consider migrating to the review-based workflow instead.", a.context.User.GetName(), req.GetUser())
```
Comment (hugoShaka) [inline lib/auth/auth_with_roles.go:2515]: > Consider migrating to the review-based workflow instead.

Do we really want to say this? One one hand we want enterprise users to fix their stuff, on the other hand we are directing OSS users to something they cannot do? Can we log this only if we're running teleport enterprise? For exmaple if we have SAML or OIDC in the modules.

Comment (fspmarshall) [reply lib/auth/auth_with_roles.go:2515]: Good point, this is worded poorly.  I'll rephrase it to something like "approval workflow" so that it can refer equally to something like `tctl request approve`/`SetAccessRequestState` or `tsh request review --approve`/`SubmitAccessReview` equally.

**Review by hugoShaka — APPROVED** [review]

**Comment by Tener** [comment]
@fspmarshall fair enough, thank you for sharing your reasoning.
